### PR TITLE
[SPARK-36133][SQL] Add empty check for catalog name

### PR DESCRIPTION
**Pull request by Peng-Lei**
### What changes were proposed in this pull request?
Add the name check when initialize a catalog by name. Keep consistent with the namespace/table naming rule.

### Why are the changes needed?
When I try to add the command `SHOW CATALOGS`, I found catalog names are meaningless in some cases. eg:
1. if the catalog name is `""`, `" "`, `"  "` that contains some spaces and so on. I think it doesn't make sense
2. if the catalog name is `"a.b"` that contains `.`, I'm more confused. `b` is the name of a namespace or part of catalog name?

[#WISH#SPARK-36133](https://issues.apache.org/jira/browse/SPARK-36133)

### Does this PR introduce _any_ user-facing change?
Yes. If user want to regist external catalog plugin. user shoule keep catalog name in SparkConf is compliant.

### How was this patch tested?
Add ut test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala [hunk 1]
```diff
@@ -44,8 +44,12 @@ class CatalogManager(
   import CatalogV2Util._
 
   private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+  private val validNameFormat = "([\\w_]+)".r
 
   def catalog(name: String): CatalogPlugin = synchronized {
+    if (!validNameFormat.pattern.matcher(name).matches()) {
+      throw QueryCompilationErrors.invalidNameForCatalogError(name)
+    }
```
Comment (Peng-Lei) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:52]: The verification of name may be a bit strict. But as an identifier. I think the basic check is needed like follow:
```scala
name.trim.isEmpty
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala [hunk 2]
```diff
@@ -44,8 +44,12 @@ class CatalogManager(
   import CatalogV2Util._
 
   private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+  private val validNameFormat = "([\\w_]+)".r
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:47]: Can we reuse the validation logic if you refer to somewhere for consistency?

Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:47]: @HyukjinKwon Thank you for taking a look at this PR firstly.  
Secondly, the validation logic for dbname/tablename is in [#L141](https://github.com/apache/spark/blob/0c34b9654187bae0638ed964efd974fc1888f46f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala#L141). But the funcation is `private`.  Also, the catalog name just from the SparkConf. it is without parse and analysis phase.
At last. I think the verification of catalog name may be a bit strict. just do a simple check [#comments](https://github.com/apache/spark/pull/33339#discussion_r669457946). Because some testcase failed eg: `org.apache.spark.sql.parquet` as a catalog name but contains `.`

**Comment by Peng-Lei** [comment]
@cloud-fan Could you take a look?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -555,6 +555,11 @@ private[spark] object QueryCompilationErrors {
       "Valid names only contain alphabet characters, numbers and _.")
   }
 
+  def invalidNameForCatalogError(name: String): Throwable = {
+    new AnalysisException(s"`$name` is not a valid name for catalog. " +
+      "Valid names can not be empty string or contains all spaces.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:560]: `... or only contain whitespaces`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 4]
```diff
@@ -799,7 +799,7 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
           }
 
           intercept[NoSuchDatabaseException] {
-            sql(s"ALTER DATABASE `db-not-exist` SET LOCATION '${tmpDir.toURI}'")
+            sql(s"ALTER DATABASE `db_not_exist` SET LOCATION '${tmpDir.toURI}'")
```
[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: I think you only add a check for catalog, why it affects how the database resolves?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 5]
```diff
@@ -1426,8 +1419,6 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
       // data source type is case insensitive
       checkAnswer(sql(s"select id from jsoN.`${f.getCanonicalPath}`"),
         df)
-      checkAnswer(sql(s"select id from `org.apache.spark.sql.json`.`${f.getCanonicalPath}`"),
-        df)
       checkAnswer(sql(s"select a.id from json.`${f.getCanonicalPath}` as a"),
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1422]: Is this change related to the catalog name validation?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MultiDatabaseSuite.scala [hunk 6]
```diff
@@ -269,46 +269,46 @@ class MultiDatabaseSuite extends QueryTest with SQLTestUtils with TestHiveSingle
   test("invalid database name and table names") {
     {
       val message = intercept[AnalysisException] {
-        df.write.format("parquet").saveAsTable("`d:b`.`t:a`")
+        df.write.format("parquet").saveAsTable("`db`.`t:a`")
       }.getMessage
-      assert(message.contains("Database 'd:b' not found"))
+      assert(message.contains("Database 'db' not found"))
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MultiDatabaseSuite.scala:274]: ditto. why need to change these tests?

**Comment by viirya** [comment]
cc @dongjoon-hyun @sunchao

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: @viirya The testcase is `ALTER DATABASE `db-not-exist` SET LOCATION '${tmpDir.toURI}` , so it will call `visitSetNamespaceLocation` first. Then `UnresolvedNamespace` to `ResolvedNamespace`, it will call the the rule `ResolveNamespace`. It goes to the last case branch ```scala
case UnresolvedNamespace(CatalogAndNamespace(catalog, ns)) =>
      ResolvedNamespace(catalog, ns)
```
In `CatalogAndNamespace` it will call `Some((catalogManager.catalog(nameParts.head), nameParts.tail))`, I add the check logic in method `catalogManager.catalog`.

In the test case , The nameParts.length == 1 and nameParts.head == `db-not-exist`, The name is not valid, so it will throw a AnalysisException. The original logic will throw a `CatalogNotFoundException`, then use `currentCatalog`, `db-not-exist` will as a name of DataBase.

Maybe I'm wrong ?

[the remark this answers is not in the document]
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: @viirya 
The testcase is `ALTER DATABASE `db-not-exist` SET LOCATION '${tmpDir.toURI}` , so it will call `visitSetNamespaceLocation` first. 
Then `UnresolvedNamespace` to `ResolvedNamespace`, it will call the the rule `ResolveNamespace`. It goes to the last case branch.

 ```scala
case UnresolvedNamespace(CatalogAndNamespace(catalog, ns)) =>
      ResolvedNamespace(catalog, ns)
```

In `CatalogAndNamespace` it will call `Some((catalogManager.catalog(nameParts.head), nameParts.tail))`, I add the check logic in method `catalogManager.catalog`.

In the test case , The nameParts.length == 1 and nameParts.head == `db-not-exist`, The name is not valid, so it will throw a AnalysisException. The original logic will throw a `CatalogNotFoundException`, then use `currentCatalog`, `db-not-exist` will as a name of DataBase.

Maybe I'm wrong ?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MultiDatabaseSuite.scala (hunk 6 above)
Comment (Peng-Lei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/MultiDatabaseSuite.scala:274]: Similar to [#comments](https://github.com/apache/spark/pull/33339#discussion_r698126897)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: Seems `validateName` was only used for validation when creating database, table before. So previously `ALTER DATABASE db-not-exist` didn't trigger the validation. It just simply checks the existing of the database `db-not-exist`.

So after this change, `ALTER DATABASE db-not-exist` will trigger catalog name check even `db-not-exist` here doesn't indicate catalog name but database name.

That being said, a command like ``ALTER DATABASE db-not-exist` will throw the `AnalysisException`, not `NoSuchDatabaseException` like before.

What happens if the database name is with catalog? As we don't validate database name for this, if it throws `NoSuchDatabaseException`, seems not consistent.

[the remark this answers is not in the document]
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: In this testcase, if there is a catalog named `db-not-exist` that same with table name `db-not-exist`, we can not find the database `db-not-exist` that it exist actually.

[the remark this answers is not in the document]
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: Maybe we can't allow catalog's name to appear in the sql statement. end-uses can switch catalogs only by using the `use` statement ?

[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: To be clear, `ALTER DATABASE db-not-exist` previously throws `NoSuchDatabaseException`, right?

After this change, as `db-not-exist` will be taken in catalog resolving `CatalogAndNamespace`, it will throw `AnalysisException`, right? So you changed the test to `db_not_exist` so it won't fail catalog name validation.

So what I mean is, now `ALTER DATABASE db-not-exist` throws `AnalysisException`, due to `db-not-exist` fails catalog name validation. But what if `ALTER DATABASE catalog.db-not-exist`? Does it still throw `NoSuchDatabaseException` like before?

If so, the two commands are pretty close or the same, as they both ask to alter a non-existing database. But one throws `AnalysisException`, and another one throws `NoSuchDatabaseException`?

I don't run it to validate. But from above description, seems it is what will happen?

[the remark this answers is not in the document]
Comment (Peng-Lei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:802]: > To be clear, `ALTER DATABASE db-not-exist` previously throws `NoSuchDatabaseException`, right?
> 
right.

> After this change, as `db-not-exist` will be taken in catalog resolving `CatalogAndNamespace`, it will throw `AnalysisException`, right? So you changed the test to `db_not_exist` so it won't fail catalog name validation.
> 
right.

> So what I mean is, now `ALTER DATABASE db-not-exist` throws `AnalysisException`, due to `db-not-exist` fails catalog name validation. But what if `ALTER DATABASE catalog.db-not-exist`? Does it still throw `NoSuchDatabaseException` like before?
> 
if `ALTER DATABASE catalog.db-not-exist`. it still throw `NoSuchDatabaseException` like before.

> If so, the two commands are pretty close or the same, as they both ask to alter a non-existing database. But one throws `AnalysisException`, and another one throws `NoSuchDatabaseException`?
> 
right.

> I don't run it to validate. But from above description, seems it is what will happen?
You are right.

**Comment by cloud-fan** [comment]
Due to the way to register a catalog right now (by configs), I feel it's a bit hard to forbid invalid catalog names. If we do want to check the name, shall we update the config setting logic, match the catalog registration configs, and validate the catalog name there? so that we can fail it earlier.

**Comment by Peng-Lei** [comment]
> Due to the way to register a catalog right now (by configs), I feel it's a bit hard to forbid invalid catalog names. If we do want to check the name, shall we update the config setting logic, match the catalog registration configs, and validate the catalog name there? so that we can fail it earlier.

Good idea. @cloud-fan  Could I do catalog name check in separate PR? In this PR just move `SessionCatalog.validateName` to a util function. So other database object that schema/database/table/view can follow the same rule easyly.

**Comment by cloud-fan** [comment]
> In this PR just move SessionCatalog.validateName to a util function.

It's only useful when we add the catalog name check, right?

**Comment by Peng-Lei** [comment]
> It's only useful when we add the catalog name check, right?

Right. I see

**Comment by Peng-Lei** [comment]
> Due to the way to register a catalog right now (by configs), I feel it's a bit hard to forbid invalid catalog names. If we do want to check the name, shall we update the config setting logic, match the catalog registration configs, and validate the catalog name there? so that we can fail it earlier.

@cloud-fan I test some testcase eg
```scala
sql("CREATE NAMESPACE IF NOT EXISTS testcat.testcat")
sql("DESCRIBE NAMESPACE testcat")
```
it will throw the exception. 
As you said: spark try to use nameParts.head to load the catalog. So if we use not a full-path and there is same name catalog, it will throw exception. Do you think this need fix?

**Comment by cloud-fan** [comment]
> it will throw the exception.

which exception does it throw? and what's the stacktrace?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala [hunk 7]
```diff
@@ -46,6 +47,7 @@ class CatalogManager(
   private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
 
   def catalog(name: String): CatalogPlugin = synchronized {
+    validateName(name)
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:50]: Instead of reusing `invalidDatabaseObjectNameError` (not sure if catalog should be considered as database object), should we have another exception such as `InvalidCatalogNameException` and throw it in `Catalogs.load`? we can also give more informative error message such as: "The catalog entry is defined and mapped to XXX but the catalog name is of invalid format" etc.

**Comment by Peng-Lei** [comment]
> > it will throw the exception.
> 
> which exception does it throw? and what's the stacktrace?

The testcase:

```scala
    withNamespace("testcat.testcat") {
      sql("CREATE NAMESPACE IF NOT EXISTS testcat.testcat COMMENT " +
        "'test namespace' LOCATION '/tmp/ns_test'")
      sql("use testcat")
      val descriptionDf = sql("DESCRIBE NAMESPACE testcat")
      ...
    }
```

The stacktrace:

```scala
next on empty iterator
java.util.NoSuchElementException: next on empty iterator
	at scala.collection.Iterator$$anon$2.next(Iterator.scala:41)
	at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
	at scala.collection.IterableLike.head(IterableLike.scala:109)
	at scala.collection.IterableLike.head$(IterableLike.scala:108)
	at scala.collection.mutable.ArrayOps$ofRef.scala$collection$IndexedSeqOptimized$$super$head(ArrayOps.scala:198)
	at scala.collection.IndexedSeqOptimized.head(IndexedSeqOptimized.scala:129)
	at scala.collection.IndexedSeqOptimized.head$(IndexedSeqOptimized.scala:129)
	at scala.collection.mutable.ArrayOps$ofRef.head(ArrayOps.scala:198)
	at scala.collection.TraversableLike.last(TraversableLike.scala:627)
	at scala.collection.TraversableLike.last$(TraversableLike.scala:626)
	at scala.collection.mutable.ArrayOps$ofRef.scala$collection$IndexedSeqOptimized$$super$last(ArrayOps.scala:198)
	at scala.collection.IndexedSeqOptimized.last(IndexedSeqOptimized.scala:135)
	at scala.collection.IndexedSeqOptimized.last$(IndexedSeqOptimized.scala:135)
	at scala.collection.mutable.ArrayOps$ofRef.last(ArrayOps.scala:198)
	at org.apache.spark.sql.execution.datasources.v2.DescribeNamespaceExec.run(DescribeNamespaceExec.scala:40)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:110)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
        ...
```

The `testcat` is a catalog of `org.apache.spark.sql.connector.catalog.InMemoryCatalog` in test package.

main cause:
when resolve Namespace the Object of `CatalogAndNamespace` 
```scala
      try {
        Some((catalogManager.catalog(nameParts.head), nameParts.tail)) // nameParts.head == "catalog" , nameParts.tail == null
      } catch {
        case _: CatalogNotFoundException =>
          Some((currentCatalog, nameParts))
      }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala (hunk 7 above)
Comment (Peng-Lei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala:50]: I think if we want keep consistent with the namespace/table naming rule, We have already considered `catalog` as a database object. And I agree with you that we should have another exception `InvalidCatalogNameException`, because the error message is more clarify.

**Comment by Peng-Lei** [comment]
> Due to the way to register a catalog right now (by configs), I feel it's a bit hard to forbid invalid catalog names. If we do want to check the name, shall we update the config setting logic, match the catalog registration configs, and validate the catalog name there? so that we can fail it earlier.

A little confused that it is reasonable that throw a invalid name exception when load a catalog(`catalog(name)`) with invalid name or `Catalogs.load` instead of use `SparkConf` to throw exception. I think maybe `SparkConf` should do common check with out `SQL` validition.  Another thing: the changes of the testcase caused by the PR are actually reasonable. 

> so that we can fail it earlier.

What's the point of fail it earlier?

**Comment by Peng-Lei** [comment]
do a sumary:
1. shall we do a validation of catalog name?
    I think it is necessary，the reason is as description of PR  [#description](https://github.com/apache/spark/pull/33339#issue-689616767)
2. do we keep the consistence with namespace/table or just empty check?
    just do empty check is simple, but it is incomplete. I considered catalog as database object, so keep the consistence with namespace/table
3. where is validation locate? `SparkConf` update or get a catalog throught `catalog(name)` or load a catalog `Catalogs.load` ?

Could you give me some advices? @cloud-fan @viirya @sunchao @HyukjinKwon @dongjoon-hyun

**Comment by cloud-fan** [comment]
> do we keep the consistence with namespace/table or just empty check?

The problem is, we can't make it consistent as different catalog implementations can define their own rules to check the namespace/table name.

**Comment by sunchao** [comment]
IMO the catalog naming convention can be enforced from Spark side. Even though I feel it doesn't necessarily need to follow the V1 database/table naming which is inherited from Hive, it should be OK so as to keep it simple and less error prone.

**Comment by Peng-Lei** [comment]
> we can't make it consistent as different catalog implementations can define their own rules to check the

Sorry, I'm late. Take a look on Presto. It do check just null and empty. I quite agree with you that we can't make it consistent as different catalog implementations can define their own rules to check the namespace/table name.

**Comment by Peng-Lei** [comment]
Update PR as follows：

1. Add empty check in `def catalog(name: String): CatalogPlugin`
2. Add `QueryCompilationErrors.invalidNameForCatalogError`

I try to add empty check in `SQLConf`, but I feel it's a bit hard to check, because the catalog name is part of SQLConf.key, not value. @cloud-fan @sunchao At last, Thank you very much for review this PR.
